src/main/java/org/hydev/mcpm/client/database/fetcher/LocalDatabaseFetcher.java

```diff
@@ -0,0 +1,192 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * A database fetcher that has a local persistent cache along and a network fallback.
+ */
+public class LocalDatabaseFetcher implements DatabaseFetcher {
+    private final URI host;
+    private final Path cacheDirectory;
+
+    private Database localDatabase;
+
+    public static final String HASH_FILE_NAME = "db.hash";
+    public static final String DATABASE_FILE_NAME = "db";
+
+    public static final String USER_AGENT = "MCPM Client";
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     * The cache directory will default to .mcpm
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     */
+    public LocalDatabaseFetcher(URI host) {
+        this(host, Path.of(".mcpm/"));
+    }
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     * @param cacheDirectory The directory that the cached database and hash file will be stored (as child files).
+     */
+    public LocalDatabaseFetcher(URI host, Path cacheDirectory) {
+        this.host = host;
+        this.cacheDirectory = cacheDirectory;
+    }
+
+    private Request requestTo(String path) {
+        return Request
+            .get(URI.create(host.toString() + "/" + path))
+            .addHeader("Host", host.getHost())
+            .addHeader("User-Agent", USER_AGENT)
+            .addHeader("Accepts", "application/json");
+    }
+
+    @Nullable
+    private String fetchHostHash() {
+        try {
+            var response = requestTo(HASH_FILE_NAME).execute();
+
+            return response.returnContent().asString().trim();
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private String fetchLocalHash() {
+        try {
+            return Files.readString(cacheDirectory.resolve(DATABASE_FILE_NAME));
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    private boolean localDatabaseIsUpToDate() {
+        var hostHash = fetchHostHash();
+        var localHash = fetchLocalHash();
+
+        // We can't reach the server right now, don't update.
+        if (hostHash == null)
+            return true;
+
+        // Missing hash, we need to update.
+        if (localHash == null)
+            return false;
+
+        return hostHash.equals(localHash);
+    }
+
+    @Nullable
+    private Database fetchLocalDatabase() {
+        try {
+            if (localDatabase != null) {
+                return localDatabase;
+            }
+
+            var file = cacheDirectory.resolve(DATABASE_FILE_NAME).toFile();
+
+            var database = new ObjectMapper().readValue(file, Database.class);
```

Comment:
Can you use the `JACKSON` constant instead of creating a new ObjectMapper every time?

Reply:
I'm not a fan of big 'Constants" files, I feel like it creates an unnecessary dependency and an extra barrier for reusing this class elsewhere. I can change it to use Constants.java if you really want, but I would rather create its own constant in-class if creating an `ObjectMapper` is expensive or keep this inline if it's not expensive.

I've moved out some of the properties of the Constants.java class into the targeting classes anyway, so (I don't think?) there should be a change in behaviour... Point me out if I'm wrong.